File: js/src/identity.ts
import { CipherSuite, type SenderContext, type Key } from 'hpke';
import { KEM_DHKEM_X25519_HKDF_SHA256, KDF_HKDF_SHA256, AEAD_AES_256_GCM } from '@panva/hpke-noble';
import { PROTOCOL, HPKE_CONFIG } from './protocol.js';
import {
  deriveResponseKeys,
  decryptChunk,
  hexToBytes,
  bytesToHex,
  HPKE_REQUEST_INFO,
  EXPORT_LABEL,
  EXPORT_LENGTH,
  RESPONSE_NONCE_LENGTH,
  ResponseKeyMaterial,
} from './derive.js';
import { ProtocolError, DecryptionError } from './errors.js';

/**
 * Request context for response decryption.
 * Holds the HPKE sender context needed to derive response keys.
 */
export interface RequestContext {
  senderContext: SenderContext;
  requestEnc: Uint8Array;
}

/**
 * Creates a new CipherSuite for X25519/HKDF-SHA256/AES-256-GCM
 */
function createSuite(): CipherSuite {
  return new CipherSuite(
    KEM_DHKEM_X25519_HKDF_SHA256,
    KDF_HKDF_SHA256,
    AEAD_AES_256_GCM
  );
}

/**
 * Identity class for managing HPKE key pairs and encryption/decryption
 */
export class Identity {
  private suite: CipherSuite;
  private publicKey: Key;
  private privateKey: Key;

  constructor(suite: CipherSuite, publicKey: Key, privateKey: Key) {
    this.suite = suite;
    this.publicKey = publicKey;
    this.privateKey = privateKey;
  }

  /**
   * Generate a new identity with X25519 key pair
   */
  static async generate(): Promise<Identity> {
    const suite = createSuite();
    const { publicKey, privateKey } = await suite.GenerateKeyPair(true); // extractable

    return new Identity(suite, publicKey, privateKey);
  }

  /**
   * Create identity from JSON string
   */
  static async fromJSON(json: string): Promise<Identity> {
    const data = JSON.parse(json);
    const suite = createSuite();

    // Deserialize keys using the suite
    const publicKey = await suite.DeserializePublicKey(new Uint8Array(data.publicKey));
    const privateKey = await suite.DeserializePrivateKey(new Uint8Array(data.privateKey), true);

    return new Identity(suite, publicKey, privateKey);
  }

  /**
   * Convert identity to JSON string
   */
  async toJSON(): Promise<string> {
    const publicKeyBytes = await this.suite.SerializePublicKey(this.publicKey);
    const privateKeyBytes = await this.suite.SerializePrivateKey(this.privateKey);

    return JSON.stringify({
      publicKey: Array.from(publicKeyBytes),
      privateKey: Array.from(privateKeyBytes),
    });
  }

  /**
   * Get public key
   */
  getPublicKey(): Key {
    return this.publicKey;
  }

  /**
   * Get public key as hex string
   */
  async getPublicKeyHex(): Promise<string> {
    const exported = await this.suite.SerializePublicKey(this.publicKey);
    return bytesToHex(exported);
  }

  /**
   * Get private key
   */
  getPrivateKey(): Key {
    return this.privateKey;
  }

  /**
   * Marshal public key configuration for server key distribution
   * Implements RFC 9458 format
   */
  async marshalConfig(): Promise<Uint8Array> {
    const kemId = HPKE_CONFIG.KEM;
    const kdfId = HPKE_CONFIG.KDF;
    const aeadId = HPKE_CONFIG.AEAD;

    // Export public key as raw bytes
    const publicKeyBytes = await this.suite.SerializePublicKey(this.publicKey);

    // Key ID (1 byte) + KEM ID (2 bytes) + Public Key + Cipher Suites
    const keyId = 0;
    const publicKeySize = publicKeyBytes.length;
    const cipherSuitesSize = 2 + 2; // KDF ID + AEAD ID

    const buffer = new Uint8Array(1 + 2 + publicKeySize + 2 + cipherSuitesSize);
    let offset = 0;

    // Key ID
    buffer[offset++] = keyId;

    // KEM ID
    buffer[offset++] = (kemId >> 8) & 0xff;
    buffer[offset++] = kemId & 0xff;

    // Public Key
    buffer.set(publicKeyBytes, offset);
    offset += publicKeySize;

    // Cipher Suites Length (2 bytes)
    buffer[offset++] = (cipherSuitesSize >> 8) & 0xff;
    buffer[offset++] = cipherSuitesSize & 0xff;

    // KDF ID
    buffer[offset++] = (kdfId >> 8) & 0xff;
    buffer[offset++] = kdfId & 0xff;

    // AEAD ID
    buffer[offset++] = (aeadId >> 8) & 0xff;
    buffer[offset++] = aeadId & 0xff;

    return buffer;
  }

  /**
   * Unmarshal public configuration from server
   */
  static async unmarshalPublicConfig(data: Uint8Array): Promise<Identity> {
    let offset = 0;

    // Read Key ID
    const keyId = data[offset++];

    // Read KEM ID
    const kemId = (data[offset++] << 8) | data[offset++];

    // Read Public Key (32 bytes for X25519)
    const publicKeySize = 32;
    const publicKeyBytes = data.slice(offset, offset + publicKeySize);
    offset += publicKeySize;

    // Read Cipher Suites Length
    const cipherSuitesLength = (data[offset++] << 8) | data[offset++];

    // Parse all cipher suites (each suite is 4 bytes: 2 for KDF, 2 for AEAD)
    const suites = [];
    const cipherSuitesEnd = offset + cipherSuitesLength;
    while (offset < cipherSuitesEnd) {
      const kdfId = (data[offset++] << 8) | data[offset++];
      const aeadId = (data[offset++] << 8) | data[offset++];
      suites.push({ kdfId, aeadId });
    }

    if (suites.length === 0) {
      throw new ProtocolError('No cipher suites found in config');
    }

    // Use the first cipher suite
    const firstSuite = suites[0];

    // Validate that we support this cipher suite
    if (firstSuite.kdfId !== HPKE_CONFIG.KDF || firstSuite.aeadId !== HPKE_CONFIG.AEAD) {
      throw new ProtocolError(
        `Unsupported cipher suite: KDF=0x${firstSuite.kdfId.toString(16)}, AEAD=0x${firstSuite.aeadId.toString(16)}`
      );
    }

    return Identity.fromPublicKeyBytes(publicKeyBytes);
  }

  /**
   * Fetch and parse the server's public identity from its HPKE key endpoint.
   *
   * This fetches without out-of-band verification. For production use,
   * make sure this is sufficient or prefer using an attested key.
   */
  static async fetchFromServer(serverURL: string): Promise<Identity> {
    const keysURL = new URL(PROTOCOL.KEYS_PATH, serverURL);
    const response = await fetch(keysURL.toString());
    if (!response.ok) {
      throw new ProtocolError(`Failed to fetch server public key: HTTP ${response.status}`);
    }
    const contentType = response.headers.get('content-type');
    if (contentType !== PROTOCOL.KEYS_MEDIA_TYPE) {
      throw new ProtocolError(
        `Invalid content type from key endpoint: expected "${PROTOCOL.KEYS_MEDIA_TYPE}", got "${contentType}"`
      );
    }
    const keysData = new Uint8Array(await response.arrayBuffer());
    return Identity.unmarshalPublicConfig(keysData);
  }

  /**
   * Create an Identity from a raw public key hex string.
   * Uses the default cipher suite (X25519/HKDF-SHA256/AES-256-GCM).
   *
   * This is used by clients who already have the server's public key
   * and don't need to fetch it.
   */
  static async fromPublicKeyHex(publicKeyHex: string): Promise<Identity> {
    const publicKeyBytes = hexToBytes(publicKeyHex);
    if (publicKeyBytes.length !== 32) {
      throw new ProtocolError(`Invalid public key length: expected 32, got ${publicKeyBytes.length}`);
    }

    return Identity.fromPublicKeyBytes(publicKeyBytes);
  }

  /**
   * Create an Identity from raw public key bytes.
   * Uses the default cipher suite (X25519/HKDF-SHA256/AES-256-GCM).
   *
   * For public-key-only identities (client-side use), we create a placeholder
   * private key that won't be used. TODO: refactor Identity to not require
   * a private key for client-side use.
   */
  private static async fromPublicKeyBytes(publicKeyBytes: Uint8Array): Promise<Identity> {
    const suite = createSuite();
    const publicKey = await suite.DeserializePublicKey(publicKeyBytes);
    const placeholderPrivateKey = await suite.DeserializePrivateKey(new Uint8Array(32), false);

    return new Identity(suite, publicKey, placeholderPrivateKey);
  }

  /**
   * Encrypt request body and return context for response decryption.
   *
   * This method is called on the SERVER's identity (public key only).
   * It:
   * 1. Creates an HPKE sender context to this identity's public key
   * 2. Encrypts the request body
   * 3. Returns a RequestContext that must be used to decrypt the response
   */
  async encryptRequestWithContext(
    request: Request
  ): Promise<{ request: Request; context: RequestContext | null }> {
    const body = await request.arrayBuffer();

    // Bodyless requests pass through unmodified - no HPKE context needed.
    // See SPEC.md Section 5.1: "When the request has no payload body, an encrypted
    // response is not possible (since there is no HPKE context to derive response
    // keys from). Such requests pass through unmodified."
    if (body.byteLength === 0) {
      return {
        request: new Request(request.url, {
          method: request.method,
          headers: request.headers,
          body: null,
        }),
        context: null,
      };
    }

    // Create sender context for encryption with info parameter for domain separation
    const infoBytes = new TextEncoder().encode(HPKE_REQUEST_INFO);
    const { encapsulatedSecret, ctx } = await this.suite.SetupSender(this.publicKey, {
      info: infoBytes,
    });

    // Store context for response decryption
    const context: RequestContext = {
      senderContext: ctx,
      requestEnc: encapsulatedSecret,
    };

    // Set headers - only encapsulated key for requests with body
    const headers = new Headers(request.headers);
    headers.set(PROTOCOL.ENCAPSULATED_KEY_HEADER, bytesToHex(context.requestEnc));

    // Encrypt the body
    const encrypted = await ctx.Seal(new Uint8Array(body));

    // Create chunked format: 4-byte length header + encrypted data
    const chunkLength = new Uint8Array(4);
    new DataView(chunkLength.buffer).setUint32(0, encrypted.byteLength, false);

    const chunkedData = new Uint8Array(4 + encrypted.byteLength);
    chunkedData.set(chunkLength, 0);
    chunkedData.set(encrypted, 4);

    return {
      request: new Request(request.url, {
        method: request.method,
        headers,
        body: chunkedData,
        duplex: 'half',
      } as RequestInit),
      context,
    };
  }

  /**
   * Decrypt response using keys derived from request context.
   *
   * This method:
   * 1. Reads the response nonce from Ehbp-Response-Nonce header
   * 2. Exports a secret from the HPKE sender context
   * 3. Derives response keys using HKDF
   * 4. Decrypts the response body
   */
  async decryptResponseWithContext(
    response: Response,
    context: RequestContext
  ): Promise<Response> {
    if (!response.body) {
      return response;
    }

    // Get response nonce from header
    const responseNonceHex = response.headers.get(PROTOCOL.RESPONSE_NONCE_HEADER);
    if (!responseNonceHex) {
      throw new ProtocolError(`Missing ${PROTOCOL.RESPONSE_NONCE_HEADER} header`);
    }

    const responseNonce = hexToBytes(responseNonceHex);
    if (responseNonce.length !== RESPONSE_NONCE_LENGTH) {
      throw new ProtocolError(
        `Invalid response nonce length: expected ${RESPONSE_NONCE_LENGTH}, got ${responseNonce.length}`
      );
    }

    // Export secret from request context
    const exportLabelBytes = new TextEncoder().encode(EXPORT_LABEL);
    const exportedSecret = await context.senderContext.Export(exportLabelBytes, EXPORT_LENGTH);

    // Derive response keys
    const km = await deriveResponseKeys(exportedSecret, context.requestEnc, responseNonce);

    // Create decrypting stream
    const decryptedStream = this.createDecryptStream(response.body, km);

    return new Response(decryptedStream, {
      status: response.status,
      statusText: response.statusText,
      headers: response.headers,
    });
  }

  /**
   * Creates a ReadableStream that decrypts response chunks.
   */
  private createDecryptStream(
    body: ReadableStream<Uint8Array>,
    km: ResponseKeyMaterial
  ): ReadableStream<Uint8Array> {
    let buffer = new Uint8Array(0);
    let seq = 0;
    const reader = body.getReader();

    return new ReadableStream({
      async pull(controller) {
        while (true) {
          // Try to read a complete chunk from buffer
          if (buffer.length >= 4) {
            const chunkLength =
              (buffer[0] << 24) | (buffer[1] << 16) | (buffer[2] << 8) | buffer[3];

            if (chunkLength === 0) {
              // Skip empty chunk
              buffer = buffer.slice(4);
              continue;
            }

            if (buffer.length >= 4 + chunkLength) {
              const ciphertext = buffer.slice(4, 4 + chunkLength);
              buffer = buffer.slice(4 + chunkLength);

              try {
                const plaintext = await decryptChunk(km, seq++, ciphertext);
                controller.enqueue(plaintext);
                return;
              } catch (error) {
                controller.error(new DecryptionError(
                  `Decryption failed at chunk ${seq - 1}`,
                  { cause: error }
                ));
                return;
              }
            }
          }

          // Need more data
          const { done, value } = await reader.read();
          if (done) {
            controller.close();
            return;
          }

          // Append to buffer
          const newBuffer = new Uint8Array(buffer.length + value.length);
          newBuffer.set(buffer);
          newBuffer.set(value, buffer.length);
          buffer = newBuffer;
        }
      },
      cancel(reason) {
        // Release the underlying reader when the stream is cancelled
        return reader.cancel(reason);
      },
    });
  }
}

File: rust/src/identity.rs
use hpke::{
    aead::AesGcm256,
    kdf::HkdfSha256,
    kem::{Kem, X25519HkdfSha256},
    setup_sender, Deserializable, OpModeS, Serializable,
};
use rand::{rngs::StdRng, SeedableRng};

use crate::{
    derive::frame_chunk,
    protocol::{
        AEAD_AES_256_GCM, EXPORT_LABEL, EXPORT_LENGTH, HPKE_REQUEST_INFO, KDF_HKDF_SHA256,
        KEM_X25519_HKDF_SHA256, KEY_ID, REQUEST_ENC_LENGTH,
    },
    session::SessionRecoveryToken,
    Error, Result,
};

type Aead = AesGcm256;
type Kdf = HkdfSha256;
type KemSuite = X25519HkdfSha256;
type PublicKey = <KemSuite as Kem>::PublicKey;

#[derive(Clone, Debug)]
pub struct ServerIdentity {
    key_id: u8,
    public_key: PublicKey,
}

impl ServerIdentity {
    pub fn from_public_key_bytes(public_key: &[u8]) -> Result<Self> {
        if public_key.len() != REQUEST_ENC_LENGTH {
            return Err(Error::InvalidConfig(format!(
                "public key must be {REQUEST_ENC_LENGTH} bytes, got {}",
                public_key.len()
            )));
        }
        let public_key = PublicKey::from_bytes(public_key)
            .map_err(|err| Error::InvalidConfig(format!("invalid X25519 public key: {err:?}")))?;
        Ok(Self {
            key_id: KEY_ID,
            public_key,
        })
    }

    pub fn from_public_key_hex(public_key_hex: &str) -> Result<Self> {
        let public_key = hex::decode(public_key_hex)?;
        Self::from_public_key_bytes(&public_key)
    }

    pub fn unmarshal_public_config(data: &[u8]) -> Result<Self> {
        let mut offset = 0usize;

        let key_id = *data
            .get(offset)
            .ok_or_else(|| Error::InvalidConfig("missing key id".into()))?;
        offset += 1;

        let kem_id = read_u16(data, &mut offset, "KEM id")?;
        if kem_id != KEM_X25519_HKDF_SHA256 {
            return Err(Error::InvalidConfig(format!(
                "unsupported KEM: 0x{kem_id:04x}"
            )));
        }

        let public_key_end = offset
            .checked_add(REQUEST_ENC_LENGTH)
            .ok_or_else(|| Error::InvalidConfig("public key offset overflow".into()))?;
        if public_key_end > data.len() {
            return Err(Error::InvalidConfig("truncated public key".into()));
        }
        let public_key_bytes = &data[offset..public_key_end];
        offset = public_key_end;

        let suites_len = read_u16(data, &mut offset, "cipher suites length")? as usize;
        if suites_len == 0 {
            return Err(Error::InvalidConfig(
                "no cipher suites found in config".into(),
            ));
        }
        if suites_len % 4 != 0 {
            return Err(Error::InvalidConfig(
                "cipher suites length must be a multiple of 4".into(),
            ));
        }
        let suites_end = offset
            .checked_add(suites_len)
            .ok_or_else(|| Error::InvalidConfig("cipher suites offset overflow".into()))?;
        if suites_end > data.len() {
            return Err(Error::InvalidConfig("truncated cipher suites".into()));
        }

        let kdf_id = read_u16(data, &mut offset, "KDF id")?;
        let aead_id = read_u16(data, &mut offset, "AEAD id")?;
        if kdf_id != KDF_HKDF_SHA256 || aead_id != AEAD_AES_256_GCM {
            return Err(Error::InvalidConfig(format!(
                "unsupported cipher suite: KDF=0x{kdf_id:04x}, AEAD=0x{aead_id:04x}"
            )));
        }

        let mut identity = Self::from_public_key_bytes(public_key_bytes)?;
        identity.key_id = key_id;
        Ok(identity)
    }

    pub fn marshal_public_config(&self) -> Vec<u8> {
        let public_key = self.public_key_bytes();
        let mut out = Vec::with_capacity(1 + 2 + public_key.len() + 2 + 4);
        out.push(self.key_id);
        out.extend_from_slice(&KEM_X25519_HKDF_SHA256.to_be_bytes());
        out.extend_from_slice(&public_key);
        out.extend_from_slice(&(4u16).to_be_bytes());
        out.extend_from_slice(&KDF_HKDF_SHA256.to_be_bytes());
        out.extend_from_slice(&AEAD_AES_256_GCM.to_be_bytes());
        out
    }

    pub fn public_key_bytes(&self) -> Vec<u8> {
        self.public_key.to_bytes().to_vec()
    }

    pub fn public_key_hex(&self) -> String {
        hex::encode(self.public_key_bytes())
    }

    pub(crate) fn encrypt_request_body(
        &self,
        plaintext: &[u8],
    ) -> Result<Option<EncryptedRequest>> {
        if plaintext.is_empty() {
            return Ok(None);
        }

        let mut csprng = StdRng::from_os_rng();
        let (enc, mut sender) = setup_sender::<Aead, Kdf, KemSuite, _>(
            &OpModeS::Base,
            &self.public_key,
            HPKE_REQUEST_INFO,
            &mut csprng,
        )
        .map_err(|err| Error::Hpke(format!("failed to set up sender: {err:?}")))?;

        let ciphertext = sender
            .seal(plaintext, &[])
            .map_err(|err| Error::Hpke(format!("failed to seal request body: {err:?}")))?;
        let body = frame_chunk(&ciphertext)?;

        let mut exported_secret = vec![0u8; EXPORT_LENGTH];
        sender
            .export(EXPORT_LABEL, &mut exported_secret)
            .map_err(|err| Error::Hpke(format!("failed to export response secret: {err:?}")))?;

        let request_enc = enc.to_bytes().to_vec();
        let token = SessionRecoveryToken::new(exported_secret, request_enc.clone())?;

        Ok(Some(EncryptedRequest {
            encapsulated_key: request_enc,
            body,
            token,
        }))
    }
}

pub(crate) struct EncryptedRequest {
    pub encapsulated_key: Vec<u8>,
    pub body: Vec<u8>,
    pub token: SessionRecoveryToken,
}

fn read_u16(data: &[u8], offset: &mut usize, field: &str) -> Result<u16> {
    if data.len().saturating_sub(*offset) < 2 {
        return Err(Error::InvalidConfig(format!("missing {field}")));
    }
    let value = u16::from_be_bytes([data[*offset], data[*offset + 1]]);
    *offset += 2;
    Ok(value)
}

#[cfg(test)]
mod tests {
    use super::*;
    use hpke::{setup_receiver, OpModeR};

    #[test]
    fn encrypts_request_body_for_hpke_receiver() {
        let mut csprng = StdRng::from_os_rng();
        let (private_key, public_key) = KemSuite::gen_keypair(&mut csprng);
        let identity = ServerIdentity {
            key_id: KEY_ID,
            public_key,
        };

        let encrypted = identity
            .encrypt_request_body(b"hello rust")
            .unwrap()
            .expect("non-empty bodies are encrypted");

        let encapped_key =
            <KemSuite as Kem>::EncappedKey::from_bytes(&encrypted.encapsulated_key).unwrap();
        let mut receiver = setup_receiver::<Aead, Kdf, KemSuite>(
            &OpModeR::Base,
            &private_key,
            &encapped_key,
            HPKE_REQUEST_INFO,
        )
        .unwrap();

        let chunk_len = u32::from_be_bytes([
            encrypted.body[0],
            encrypted.body[1],
            encrypted.body[2],
            encrypted.body[3],
        ]) as usize;
        let plaintext = receiver
            .open(&encrypted.body[4..4 + chunk_len], &[])
            .unwrap();

        let mut exported_secret = vec![0u8; EXPORT_LENGTH];
        receiver.export(EXPORT_LABEL, &mut exported_secret).unwrap();

        assert_eq!(plaintext, b"hello rust");
        assert_eq!(encrypted.token.exported_secret, exported_secret);
        assert_eq!(encrypted.token.request_enc, encrypted.encapsulated_key);
    }

    #[test]
    fn empty_request_body_is_plaintext() {
        let mut csprng = StdRng::from_os_rng();
        let (_, public_key) = KemSuite::gen_keypair(&mut csprng);
        let identity = ServerIdentity {
            key_id: KEY_ID,
            public_key,
        };

        assert!(identity.encrypt_request_body(b"").unwrap().is_none());
    }
}
